### ISSUE TITLE
[BUG]Stack overflow: *** stack smashing detected ***

### ISSUE THREAD

**Issue by sab24** [RRZE-HPC/likwid#334]
Compiler warns about overflow when compiling with make using default settings and GCC 10.2
```
===>  COMPILE  GCC/thermal.o
===>  COMPILE  GCC/timer.o
===>  COMPILE  GCC/topology.o
./src/topology.c: In function ‘readTopologyFile’:
./src/topology.c:444:50: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  444 |                     cpuid_info.osname[maxStrLen] = '\0';
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
./src/topology.c:436:45: note: at offset 257 to an object with size 0 allocated by ‘malloc’ here
  436 |                 cpuid_info.osname = (char*) malloc(maxStrLen * sizeof(char));
      |                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

And also
```
===>  COMPILE  GCC/loadData.o
===>  ENTER  /tmp/likwid-5.0.2/ext/hwloc
===>  ENTER  /tmp/likwid-5.0.2/ext/lua
In function ‘createstrobj’,
    inlined from ‘luaS_createlngstrobj’ at ./src/lstring.c:148:17,
    inlined from ‘luaS_newlstr’ at ./src/lstring.c:206:10:
./src/lstring.c:142:17: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  142 |   getstr(ts)[l] = '\0';  /* ending 0 */
In file included from ./includes/ldebug.h:11,
                 from ./src/lstring.c:17:
./src/lstring.c: In function ‘luaS_newlstr’:
./includes/lstate.h:185:18: note: at offset 0 to object ‘ts’ with size 24 declared here
  185 |   struct TString ts;
      |                  ^~
===>  CREATE SHARED LIB  liblikwid.so
===>  CREATE LIB  liblikwidpin.so

```

patching the file src/toplogy.c to 
```
                    cpuid_info.osname[maxStrLen -1] = '\0';
```

gives
```
 likwid-topology
/usr/local/bin/likwid-lua: error loading module 'liblikwid' from file '/usr/local/share/lua/5.3/liblikwid.lua':
        /usr/local/share/lua/5.3/liblikwid.lua:330: too many C levels (limit is 200) in function at line 313 near 'i'
stack traceback:
        [C]: in ?
        [C]: in function 'require'
        /usr/local/share/lua/5.3/liblikwid.lua:33: in main chunk
        [C]: in function 'require'
        /usr/local/share/lua/5.3/liblikwid.lua:33: in main chunk
        [C]: in function 'require'
        /usr/local/share/lua/5.3/liblikwid.lua:33: in main chunk
        [C]: in function 'require'
        /usr/local/share/lua/5.3/liblikwid.lua:33: in main chunk
        [C]: in function 'require'
        ...
        [C]: in function 'require'
        /usr/local/share/lua/5.3/liblikwid.lua:33: in main chunk
        [C]: in function 'require'
        /usr/local/share/lua/5.3/liblikwid.lua:33: in main chunk
        [C]: in function 'require'
        /usr/local/share/lua/5.3/liblikwid.lua:33: in main chunk
        [C]: in function 'require'
        /usr/local/share/lua/likwid.lua:33: in main chunk
        [C]: in function 'require'
        /usr/local/bin/likwid-topology:35: in main chunk
        [C]: in ?

```
This is with version 5.0.2,  archlinux X86-64 on AMD Zen1

**Comment by TomTheBear** [comment]
Thanks for reporting. I should add newest compiler versions to LIKWID's CI actions.

The first problem in `src/topology.c` is already fixed in the master branch.

The second warning is inside Lua, so out of scope of LIKWID. I checked the latest Lua releases 5.3.6 and 5.4.1 and only 5.4.1 does no throw the warning. I checked the main changes from 5.3 to 5.4 and none should affect LIKWID's Lua scripts, but I'll do some further tests.

**Comment by sab24** [comment]
Thanks for the quick reply. Yes at first I checkout out head and since there was an error I used the latest release version. Indeed the topology overflow is fixed in master branch, but I still have a stack overflow:
```
 lldb /usr/local/bin/likwid-lua
(lldb) target create "/usr/local/bin/likwid-lua"
Current executable set to '/usr/local/bin/likwid-lua' (x86_64).
(lldb) r /usr/local/bin/likwid-topology
Process 235338 launched: '/usr/local/bin/likwid-lua' (x86_64)
Process 235338 stopped and restarted: thread 1 received signal: SIGCHLD
*** stack smashing detected ***: terminated
Process 235338 stopped
* thread #1, name = 'likwid-lua', stop reason = signal SIGABRT
    frame #0: 0x00007ffff7c89615 libc.so.6`raise + 325
libc.so.6`raise:
->  0x7ffff7c89615 <+325>: movq   0x108(%rsp), %rax
    0x7ffff7c8961d <+333>: subq   %fs:0x28, %rax
    0x7ffff7c89626 <+342>: jne    0x7ffff7c8964c            ; <+380>
    0x7ffff7c89628 <+344>: movl   %r8d, %eax
(lldb) bt
* thread #1, name = 'likwid-lua', stop reason = signal SIGABRT
  * frame #0: 0x00007ffff7c89615 libc.so.6`raise + 325
    frame #1: 0x00007ffff7c72862 libc.so.6`abort + 278
    frame #2: 0x00007ffff7ccb5e8 libc.so.6`__libc_message + 616
    frame #3: 0x00007ffff7d5ba3a libc.so.6`__fortify_fail + 42
    frame #4: 0x00007ffff7d5ba04 libc.so.6`__stack_chk_fail + 20
    frame #5: 0x00007ffff70bacdb liblikwid.so`hwloc_init_nodeTopology + 1275
```

**Comment by TomTheBear** [comment]
This seems to be in `src/topology_hwloc.c` but I'm not sure where exactly. `hwloc_init_nodeTopology + 1275` is not an easy-to-find location specification. Maybe you rebuild LIKWID (`make distclean && make`) with `DEBUG=true` in `config.mk` so we get better location data.

I can run `likwid-topology` with my GCC 10.2 build without a problem. I have an idea where the smashing could happen but I'm not sure and it would be strange somehow (_"your CPU has 8 cores but one core has an OS id < 0 or > 7"_).

**Comment by sab24** [comment]
It's at the end of the function 


```
    0x7ffff709438c <+2726>: movl   %edx, 0xc(%rax)
    0x7ffff709438f <+2729>: movl   -0x88(%rbp), %edx
    0x7ffff7094395 <+2735>: leaq   0xbb55e4(%rip), %rax      ; cpuid_topology
    0x7ffff709439c <+2742>: movl   %edx, 0x8(%rax)
    0x7ffff709439f <+2745>: nop
    0x7ffff70943a0 <+2746>: movq   -0x8(%rbp), %rax
    0x7ffff70943a4 <+2750>: subq   %fs:0x28, %rax
    0x7ffff70943ad <+2759>: je     0x7ffff70943b4            ; <+2766> at topology_hwloc.c:557:1
    0x7ffff70943af <+2761>: callq  0x7ffff6fd3440            ; symbol stub for: __stack_chk_fail
->  0x7ffff70943b4 <+2766>: leave
    0x7ffff70943b5 <+2767>: retq
 bt
* thread #1, name = 'likwid-lua', stop reason = signal SIGABRT
    frame #0: 0x00007ffff7c89615 libc.so.6`raise + 325
    frame #1: 0x00007ffff7c72862 libc.so.6`abort + 278
    frame #2: 0x00007ffff7ccb5e8 libc.so.6`__libc_message + 616
    frame #3: 0x00007ffff7d5ba3a libc.so.6`__fortify_fail + 42
    frame #4: 0x00007ffff7d5ba04 libc.so.6`__stack_chk_fail + 20
  * frame #5: 0x00007ffff70943b4 liblikwid.so`hwloc_init_nodeTopology(cpuSet=cpu_set_t @ 0x00007fffffffc7d0) at topology_hwloc.c:557:1
(lldb) list
   504          if (obj->type == HWLOC_OBJ_SOCKET)
   505          {
   506              hwThreadPool[id].packageId = obj->os_index;
   507          }
   508          else
   509          {
   510              hwThreadPool[id].packageId = 0;
   511          }
   512  #else
   513          hwThreadPool[id].packageId = obj->os_index;

```


My CPU has 32 cores, would that be the problem?

**Comment by TomTheBear** [comment]
No, 32 cores is commonly not a problem. The limit for CPU cores is currently `MAX_NUM_THREADS=300` but is configurable. I ran LIKWID already on nodes with more than 1000 CPU cores (128 CPU sockets). Is SMT on or off? Because there is this part (https://github.com/RRZE-HPC/likwid/blob/master/src/topology_hwloc.c#L470-L488) which is only executed by AMD CPUs and uses `id-1` which could be out of bounds. I don't have an AMD system with GCC 10.2 available at the moment.

The debugging output is somehow mixed. Line 557 in `src/topology_hwloc.c` is the closing curly bracket. But `list` shows completely irrelevant lines.

Another problematic location could be https://github.com/RRZE-HPC/likwid/blob/master/src/topology_hwloc.c#L523. Maybe you replace the `16` with `MAX_NUM_NODES` (=64, configurable in `config.mk`). See https://github.com/RRZE-HPC/likwid/commit/c9925a67648e7effa9b7fa36e0726a10ff845440 .

**Comment by sab24** [comment]
Good to know high core cpus are supported. MAX_NUM_NODES is already 64 with the git version I am using now. I checked   /sys/devices/system/cpu/smt/active and it gives 1 on the bare-metal system, but I am running from within KVM where it is 0. In KVM there are 60 physical sockets with 1 core each, but the real processor has 32 cores with 2 threads per core. Maybe this is why it crashes?

Ah  yes, list shows a different output as I was playing with it earlier, decrementing the line count. The error occurs where there is a return from that function. This shows it better:
```
(lldb) list 557
   557  }
   558
   559  void hwloc_split_llc_check(CacheLevel* llc_cache)
   560  {
   561      int i = 0;
   562      hwloc_obj_t obj = NULL;
   563      int num_sockets = likwid_hwloc_get_nbobjs_by_type(hwloc_topology, HWLOC_OBJ_SOCKET);
   564      int num_nodes = likwid_hwloc_get_nbobjs_by_type(hwloc_topology, HWLOC_OBJ_NODE);
   565      if (num_sockets == num_nodes)
   566      {
(lldb) list -10
   547              {
   548                  hwThreadPool[i].packageId = j;
   549              }
   550          }
   551      }
   552      cpuid_topology.threadPool = hwThreadPool;
   553      cpuid_topology.numThreadsPerCore = maxNumLogicalProcsPerCore;
   554      cpuid_topology.numCoresPerSocket = maxNumCoresPerSocket;
   555      cpuid_topology.numSockets = maxNumSockets;
   556      return;
(lldb) e maxNumSockets
(int) $0 = 60
(lldb) e cpuid_topology
(CpuTopology) $1 = {
  numHWThreads = 60
  activeHWThreads = 60
  numSockets = 60
  numCoresPerSocket = 1
  numThreadsPerCore = 1
  numCacheLevels = 0
  threadPool = 0x00005555555fe0e0
  cacheLevels = 0x0000000000000000
  topologyTree = 0x0000000000000000
}
```

**Comment by TomTheBear** [comment]
Have you done a `git pull && make distclean && make` to get the [latest commit](https://github.com/RRZE-HPC/likwid/commit/c9925a67648e7effa9b7fa36e0726a10ff845440)? Because otherwise the code is limited to 16 sockets independent of the value in `MAX_NUM_NODES` (despite the values in cpuid_topology look reasonable).

I cannot recommend the usage of LIKWID inside virtual environments. There might be combinations of guest os, host os and visualization solutions that work but LIKWID was thought to run on bare-metal. The topology, pinning and benchmarking stuff probably works but for performance monitoring stuff is problematic in virtual environments.

**Comment by sab24** [comment]
Yeah, did a clean rebuild. Also ran make uninstall and removed everything. Currently I don’t have any other machine available than the VM and ARM machines (which fail to compile). Maybe I can setup PXE to do a test run on intel based systems and later debug the assembly side to side with a working version to see why it crashes.

**Comment by TomTheBear** [comment]
The latest commit adds also a guard to not write out-of-bounds of `hwThreadPool` (which is also allocated on the heap, not stack). The only stack allocation in this function is the `socket_nums` array. You could run it with `-V 3` on the command line to see whether each HW thread is detected properly.

For ARM builds, you have to change the `COMPILER` setting in `config.mk` to `GCCARMv8` (or `GCCARMv7`).

**Comment by sab24** [comment]
Works, somehow the last commit changed something and now the error is gone on AMD64.

For ARM64 I changed the config.mk to GCCARMv7  but still have errors:
```
===>  COMPILE  GCCARMv7/thermal.o
===>  COMPILE  GCCARMv7/timer.o
===>  COMPILE  GCCARMv7/topology.o
===>  COMPILE  GCCARMv7/topology_hwloc.o
./src/topology_hwloc.c: In function 'hwloc_init_cpuInfo':
./src/topology_hwloc.c:344:9: error: 'count' undeclared (first use in this function)
  344 |     if (count > cpuid_topology.numHWThreads)
      |         ^~~~~
./src/topology_hwloc.c:344:9: note: each undeclared identifier is reported only once for each function it appears in

```

**Comment by TomTheBear** [comment]
Great that it works for AMD64 now.

Are you sure you need `GCCARMv7` and not `GCCARMv8`? Nevertheless, I fixed the ARMv7 error with the last commit.

**Comment by sab24** [comment]
Thanks for the fix. I guess I need GCCARMv7 as it is a raspberry pi 4 with 32bit operating system, still using the features of the armv7 architecture and the flag -mabi=lp64 is not working on this os.  I now have a new error:
```
===>  COMPILE  GCCARMv7/pci_proc.o
===>  COMPILE  GCCARMv7/perfgroup.o
===>  COMPILE  GCCARMv7/perfmon.o
In file included from ./src/perfmon.c:84:
./src/includes/perfmon_perfevent.h: In function 'perf_pmc_setup':
./src/includes/perfmon_perfevent.h:336:64: error: 'A64FX_FX1000' undeclared (first use in this function)
  336 |     if (cpuid_info.vendor == FUJITSU_ARM && cpuid_info.part == A64FX_FX1000)
      |                                                                ^~~~~~~~~~~~
./src/includes/perfmon_perfevent.h:336:64: note: each undeclared identifier is reported only once for each function it appears in
make: *** [Makefile:277: GCCARMv7/perfmon.o] Error 1
```

Also I saw in the code there was a release date of 23rd of December 2019, that was not updated during release 5.0.2 as it still points to 5.0.1.

**Comment by TomTheBear** [comment]
Fixed it. I renamed `A64FX_FX1000` to `FUJITSU_A64FX` and forgot this file.

**Comment by sab24** [comment]
Did a reinstall today to get both armv7 and armv8 working and both compile and run! I hope I helped making this project a bit better and thanks for the good support even during evenings and weekends.
